spec/lucky/action_spec.cr

```diff
@@ -146,6 +157,49 @@ describe Lucky::Action do
       action.params.get(:q).should eq "test"
     end
   end
+
+  describe "optional params" do
+    it "is initialized to nil" do
+      action = OptionalParams::Index.new(build_context(path: ""), params)
+      action.page.should eq nil
+    end
+
+    it "is fetched if present" do
+      action = OptionalParams::Index.new(build_context(path: "/?page=3"), params)
+      action.page.should eq 3
+    end
+
+    it "can be used within the action" do
+      response = OptionalParams::Index.new(build_context(path: "/?page=3"), params).call
+      response.body.to_s.should eq "optional param: 3 1 1337"
+    end
+
+    it "can specify a default value" do
+      action = OptionalParams::Index.new(build_context(path: ""), params)
+      action.with_default.should eq "default"
+    end
+
+    it "is added as optional argument to the path" do
+      OptionalParams::Index.path(7).should eq "/optional_params?page=7"
+      OptionalParams::Index.path(7, "/other").should eq "/optional_params?page=7&with_default=%2Fother"
+    end
+
+    it "is added as optional argument to the route" do
+      OptionalParams::Index.route(7).should eq Lucky::RouteHelper.new(:get, "/optional_params?page=7")
+      OptionalParams::Index.route(7, "/other").should eq Lucky::RouteHelper.new(:get, "/optional_params?page=7&with_default=%2Fother")
+    end
+
+    it "doesnt raise when we cannot parse the optional param into the desired type" do
+      response = OptionalParams::Index.new(build_context(path: "/?page=no_int"), params).call
+      response.body.to_s.should eq "optional param:  1 1337"
+    end
+
+    it "raises when we cannot parse the non-optional param into the desired type" do
```

Comment:
This is bollocks actually I don't think we ever want to raise/runtime error when something inside the query params cannot be parsed. We could just return the default value instead duh.

Reply:
I think that if a parameter can’t be parsed then that is an error that should be raised. I think it would be a good idea to make a custom error that we can raise so it can be handled if you want. 

Something like `Lucky::InvalidParamError`. It should probably accept (in the initializer) the param name, and the value that it tried to parse, and the type it tried to parse it to

Reply:
Yep agreed I extracted two custom Exceptions. Feel free to comment on naming/positioning ✅